### DISCUSSION TITLE
Background color for math in display mode: Rendering a zero-padding box?

### DISCUSSION THREAD

**Discussion by GiacoCorsiglia** [KaTeX/KaTeX#3751] [Q&A]
I'd like to introduce a red background only behind the $\frac{a}{b}$ in the below equation:

$$2 \frac{a}{b}$$

The $2$ should not have a background color.

I naively tried to accomplish this via:

```tex
2 \htmlClass{red-background}{\frac{a}{b}}
```

but that produces a pretty weird result:
<img width="71" alt="image" src="https://user-images.githubusercontent.com/15148404/202622808-6568c819-ca73-41c1-95c5-95f5433803f4.png">

I'm wondering if there's a way got get a zero-padding box around an element.  Then I could do

```tex
2 \htmlClass{red-background}{\zeroPaddingBox{\frac{a}{b}}}
```

and presumably my CSS would now work.

One thing that's important is I would like the spacing to be exactly the same whether the box is there or not.

I know about `\colorbox`, but it switches its contents to text mode and I wasn't sure how to get them back into _display_ math mode.  It also adds extra spacing around the element.

Any suggestions?  Thanks!

**Comment by mbourne** [comment]
Giaco

In case there is no native solution and assuming you're open to some post-processing, you could do this.

Add the following styles:

```
.mfrac{position:relative}
.red-box{background:#fee;position:absolute;left:-2px;border:1px solid red}
```

Then add this script:

```
window.addEventListener('load', function() {

	var ks = document.querySelectorAll('.katex');
	
	ks.forEach(function(k) {
		var t = k.querySelector('.base').offsetTop - k.querySelector('.mfrac').offsetTop;
		var w = k.querySelector('.mfrac .vlist-t2').offsetWidth;
		var h = k.querySelector('.mfrac .vlist-t2').offsetHeight;
		var span = document.createElement('span');
		span.classList.add('red-box');
		span.style.width = w+"px";
		span.style.height = (h+5)+"px";
		span.style.top = (t-4)+"px";
		var mfrac = k.querySelector('.mfrac');
		mfrac.insertBefore(span, mfrac.firstChild);
	});
});
```
The script gets the required height for the new box and its position relative to the wrapping span.mfrac, then inserts the new span behind the fraction. I'm assuming you'll have more than one fraction on the page, and this will put a box around the first fraction in each math expression.

Result for 3 different-size fractions:

![image](https://user-images.githubusercontent.com/7541940/202820940-977da093-fb91-4c7b-a260-5968912becf6.png)

It adds no extra spacing.

Hope it helps. Regards,
Murray

**Reply by GiacoCorsiglia** [reply]
Thanks a bunch @mbourne!  I'm going to leave this question open in hopes of a native solution (and one that works for more than just fractions), but post-processing is definitely a good idea, so this is super helpful.

**Comment by edemaine** [comment]
> I know about `\colorbox`, but it switches its contents to text mode and I wasn't sure how to get them back into display math mode.

Here's how you work around that:

```latex
2 \colorbox{red}{$\displaystyle \frac{a}{b}$}
```

> It also adds extra spacing around the element.

This is trickier. Reading the LaTeX source code (specifically `xcolor.sty` which defines `\colorbox`), it looks like the "right" way to configure this is to `\setlength{\fboxsep}{0pt}`. Unfortunately, KaTeX doesn't yet support `\setlength` (#687).

One way to change this globally is to override the CSS rule:

```css
.katex .boxpad { padding: 0 }
```

Example rendering:

![image](https://user-images.githubusercontent.com/2218736/202857201-4efbfaee-4be9-4128-9ff0-1f4c73eaee24.png)

The alternative is to manually undo the spacing in KaTeX, which hard-codes a -3pt (0.3em) space.  Example:

```latex
2 \frac{a}{b}
\\[1ex]
2 \colorbox{red}{\hspace{-3pt}$\displaystyle \frac{a}{b}$\hspace{-3pt}}
```

![image](https://user-images.githubusercontent.com/2218736/202857566-f8ba4aa1-af18-4c5a-8088-55696ab9d93d.png)

You can make this into a macro for easier use:

```latex
\gdef\red#1{\colorbox{red}{\hspace{-3pt}$\displaystyle #1$\hspace{-3pt}}}
2 \red{\frac{a}{b}}
```
